src/Data/Singletons/Prelude/Monad/Internal.hs

```diff
@@ -0,0 +1,511 @@
+      -- The default definition is @'fmap' . 'const'@, but this may be
+      -- overridden with a more efficient version.
+      (<$)        :: a -> f b -> f a
+      (<$)        =  fmap . const
+
+  infixl 4 <*>, <*, *>, <**>
+
+  -- -| A functor with application, providing operations to
+  --
+  -- -* embed pure expressions ('pure'), and
+  --
+  -- -* sequence computations and combine their results ('<*>' and 'liftA2').
+  --
+  -- A minimal complete definition must include implementations of 'pure'
+  -- and of either '<*>' or 'liftA2'. If it defines both, then they must behave
+  -- the same as their default definitions:
+  --
+  --      @('<*>') = 'liftA2' 'id'@
+  --
+  --      @'liftA2' f x y = f '<$>' x '<*>' y@
+  --
+  -- Further, any definition must satisfy the following:
+  --
+  -- [/identity/]
+  --
+  --      @'pure' 'id' '<*>' v = v@
+  --
+  -- [/composition/]
+  --
+  --      @'pure' (.) '<*>' u '<*>' v '<*>' w = u '<*>' (v '<*>' w)@
+  --
+  -- [/homomorphism/]
+  --
+  --      @'pure' f '<*>' 'pure' x = 'pure' (f x)@
+  --
+  -- [/interchange/]
+  --
+  --      @u '<*>' 'pure' y = 'pure' ('$' y) '<*>' u@
+  --
+  --
+  -- The other methods have the following default definitions, which may
+  -- be overridden with equivalent specialized implementations:
+  --
+  --   * @u '*>' v = ('id' '<$' u) '<*>' v@
+  --
+  --   * @u '<*' v = 'liftA2' 'const' u v@
+  --
+  -- As a consequence of these laws, the 'Functor' instance for @f@ will satisfy
+  --
+  --   * @'fmap' f x = 'pure' f '<*>' x@
+  --
+  --
+  -- It may be useful to note that supposing
+  --
+  --      @forall x y. p (q x y) = f x . g y@
+  --
+  -- it follows from the above that
+  --
+  --      @'liftA2' p ('liftA2' q u v) = 'liftA2' f u . 'liftA2' g v@
+  --
+  --
+  -- If @f@ is also a 'Monad', it should satisfy
+  --
+  --   * @'pure' = 'return'@
+  --
+  --   * @('<*>') = 'ap'@
+  --
+  --   * @('*>') = ('>>')@
+  --
+  -- (which implies that 'pure' and '<*>' satisfy the applicative functor laws).
+
+  class Functor f => Applicative (f :: Type -> Type) where
+      -- {-# MINIMAL pure, ((<*>) | liftA2) #-}
+      -- -| Lift a value.
+      pure :: a -> f a
+
+      -- -| Sequential application.
+      --
+      -- A few functors support an implementation of '<*>' that is more
+      -- efficient than the default one.
+      (<*>) :: f (a -> b) -> f a -> f b
+      (<*>) = liftA2 id
+
+      -- -| Lift a binary function to actions.
+      --
+      -- Some functors support an implementation of 'liftA2' that is more
+      -- efficient than the default one. In particular, if 'fmap' is an
+      -- expensive operation, it is likely better to use 'liftA2' than to
+      -- 'fmap' over the structure and then use '<*>'.
+      liftA2 :: (a -> b -> c) -> f a -> f b -> f c
+      liftA2 f x = (<*>) (fmap f x)
+
+      -- -| Sequence actions, discarding the value of the first argument.
+      (*>) :: (f :: Type -> Type) a -> f b -> f b
```

Comment:
We should document why `f` needs a kind signature here.

Reply:
I've added a lengthy `Note [How to get the right kinds when promoting Functor and friends]` to explain the subtleties involved.